TrackletAlgorithm/MatchProcessor.h

```diff
@@ -0,0 +1,903 @@
+//-----------------------------------------------------------------------------------------------------------
+
+// MatchCalculator
+namespace MC {
+  enum imc {UNDEF_ITC, A = 0, B = 1, C = 2, D = 3, E = 4, F = 5, G = 6, H = 7, I = 8, J = 9, K = 10, L = 11, M = 12, N = 13, O = 14};
+}
+
+template<regionType ASTYPE, regionType APTYPE, regionType VMSMEType, regionType FMTYPE, int maxFullMatchCopies, int LAYER=TF::L1, MC::imc PHISEC=MC::A>
+void MatchCalculator(BXType bx,
+                     ap_uint<1> newtracklet,
+                     ap_uint<1>& savedMatch,
+                     ap_uint<17>& best_delta_phi,
+                     const AllStubMemory<ASTYPE>* allstub,
+                     const AllProjection<APTYPE>& proj,
+                     ap_uint<VMProjectionBase<BARREL>::kVMProjIndexSize> projid,
+                     ap_uint<VMStubMECMBase<VMSMEType>::kVMSMEIDSize> stubid,
+                     BXType& bx_o,
+                     int &nmcout1,
+                     int &nmcout2,
+                     int &nmcout3,
+                     int &nmcout4,
+                     int &nmcout5,
+                     int &nmcout6,
+                     int &nmcout7,
+                     int &nmcout8,
+                     FullMatchMemory<BARREL> fullmatch[maxFullMatchCopies]
+){
+
+#pragma HLS inline
+
+  using namespace PR;
+
+
+  // Setup constants depending on which layer/disk working on
+  // probably should move these to constants file
+  const ap_uint<4> kNbitszprojL123 = 12; // nbitszprojL123 in emulation (defined in constants) 
+  const ap_uint<4> kNbitszprojL456 = 8;  // nbitszprojL456 in emulation (defined in constants)
+  const ap_uint<5> kNbitsdrinv = 19;     // idrinvbits     in emulation (defined in constants)
+  const ap_uint<4> kShift_Rinv = 13;     // rinvbitshift   in emulation (defined in constants)
+  const ap_uint<3> kShift_Phider = 7;    // phiderbitshift in emulation (defined in constants)
+  const ap_uint<3> kNbitsrL123 = 7;      // nbitsrL123     in emulation (defined in constants)
+  const ap_uint<3> kNbitsrL456 = 7;      // nbitsrL456     in emulation (defined in constants) 
+  const ap_int<4>  kShift_PS_zderL = -7; // PS_zderL_shift in emulation (defined in constants)
+  const ap_int<4>  kShift_2S_zderL = -7; // SS_zderL_shift in emulation (defined in constants)
+
+  const auto kFact               = (LAYER < TF::L4)? 1 : (1<<(kNbitszprojL123-kNbitszprojL456)); // fact_ in emulation defined in MC
+  const auto kPhi0_shift         = (LAYER < TF::L4)? 3 : 0;                                      // phi0shift_ in emulation defined in MC
+  constexpr int  kShift_phi0bit      = 1;                                                             // phi0bitshift in emulation defined in constants
+  const ap_uint<10> kPhi_corr_shift_L123 = 7 + kNbitsdrinv + kShift_phi0bit - kShift_Rinv - kShift_Phider;                    // icorrshift for L123
```

Comment:
If you're bored waiting for the tarball update, you could always try and create some descriptive constants for the hardcoded numbers. I currently have no idea where the 10 and 7 comes from here for example.

Reply:
This looks like a "todo" item in the emulation as well.
https://github.com/cms-L1TK/cmssw/blob/37e2596707745f16927215871e701a46edf3b74d/L1Trigger/TrackFindingTracklet/src/MatchCalculator.cc#L34-L50